File: infra/src/scripts/internal/src/internal/database/models/datasets.py
from internal.utilities.parser import make_id
from pydantic import BaseModel, computed_field, Field
from datetime import datetime

from .shared import ModalityTypeEnum


class DatasetModel(BaseModel):
    id: str
    name: str
    description: str
    modality: ModalityTypeEnum
    batch_size: int
    batch_keys: list[str] = []
    created_at: datetime
    updated_at: datetime

    @property
    def has_batches(self) -> bool:
        return len(self.batch_keys) != 0


class DatasetCreateModel(BaseModel):
    name: str
    description: str
    modality: ModalityTypeEnum
    batch_size: int

    @computed_field
    @property
    def id(self) -> str:
        return make_id(self.name)

    @computed_field
    @property
    def created_at(self) -> datetime:
        return datetime.now()

    @computed_field
    @property
    def updated_at(self) -> datetime:
        return datetime.now()


class DatasetBatchDownloadCreateModel(BaseModel):
    batch_key: str

File: web_app/lib/models/datasets.ts
import { z } from "zod";
import { ModalityTypeEnum } from "./shared";

export const DatasetModelSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
  modality: ModalityTypeEnum,
  batch_size: z.number(),
  batch_keys: z.array(z.string()).default([]),
  created_at: z.string(),
  updated_at: z.string(),
  is_completed: z.boolean().default(false),
  is_deleted: z.boolean().default(false),
});

export type DatasetModel = z.infer<typeof DatasetModelSchema>;

export const DatasetCreateModelSchema = z.object({
  name: z.string().min(1, "Name is required"),
  description: z.string().min(1, "Description is required"),
  modality: ModalityTypeEnum,
  batch_size: z.coerce.number().min(1, "Batch size must be at least 1"),
});

export type DatasetCreateModel = z.infer<typeof DatasetCreateModelSchema>;
